### PR TITLE
E2E with HCP Consul/Vault

### PR DESCRIPTION
**Pull request by tgross**
Use HCP Consul and HCP Vault for the Consul and Vault clusters used in E2E testing. This has the following benefits:

* Without the need to support mTLS bootstrapping for Consul and Vault, we can simplify the mTLS configuration by leaning on Terraform instead of janky bash shell scripting.
* Vault bootstrapping is no longer required, so we can eliminate even more janky shell scripting
* Our E2E exercises HCP, which is important to us as an organization
* With the reduction in configurability, we can simplify the Terraform configuration and drop the complicated `provision.sh`/`provision.ps1` scripts we were using previously. We can template Nomad configuration files and upload them with the `file` provisioner.
* Packer builds for Linux and Windows become much simpler.

tl;dr way less janky shell scripting!

At this point I've tested out most of the E2E suite with this configuration, in particular the `Connect` and `VaultSecrets` suite that are the most likely to break. I still need to complete full testing. There was a parallel PR merged in the `nomad-e2e` repo to set up the infrastructure.

Follow-up issues:
* https://github.com/hashicorp/nomad/issues/12314
* https://github.com/hashicorp/nomad/issues/12315
* https://github.com/hashicorp/nomad/issues/12316

There will be a PR in the ENT repo for the ENT-only Consul namespaces test as well.

Tests passing against a current build of `main` from the Nomad ENT repo (plus a couple patches from this PR):
- [x] affinities
- [x] connect/Connect 
- [x] connect/ConnectACLs
- [x] consul/Consul
- [x] consul/ConsulNamespaces
- [x] consultemplate
- [x] csi
- [x] deployment
- [x] eval_priority
- [x] events
- [x] simple
- [x] isolation
- [x] lifecycle
- [x] metrics
- [x] namespaces
- [x] networking
- [x] nodedrain
- [x] nomadexec
- [x] oversubscription
- [x] parameterized
- [x] periodic
- [x] quotas
- [x] rescheduling
- [x] scaling
- [x] scalingpolicies
- [x] scheduler_sysbatch
- [x] scheduler_system
- [x] spread
- [x] taskevents
- [x] vaultsecrets
- [x] volumes

### CODE REVIEW

**Review by shoenig — APPROVED** [review]
LGTM! My only real question is about the file permissions and maybe the consul user. 

Thanks for doing this!

File: e2e/terraform/hcp_consul.tf [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+data "consul_acl_token_secret_id" "consul_agent_token" {
+  accessor_id = consul_acl_token.consul_agent_token.id
+}
+
+resource "local_file" "consul_acl_file" {
+  sensitive_content = templatefile("etc/consul.d/client_acl.json", {
+    token = data.consul_acl_token_secret_id.consul_agent_token.secret_id
+  })
+  filename        = "uploads/shared/consul.d/client_acl.json"
+  file_permission = "0700"
```
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/terraform/hcp_consul.tf]: `0600` should work. Pushed that and will test.

File: e2e/terraform/etc/consul.d/consul.service [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+After=network-online.target
+
+[Service]
+Restart=on-failure
+Environment=CONSUL_ALLOW_PRIVILEGED_PORTS=true
+WorkingDirectory=/etc/consul.d
+ExecStart=/usr/bin/consul agent -config-dir="/etc/consul.d"
+ExecReload=/bin/kill -HUP $MAINPID
+KillSignal=SIGTERM
+User=root
```
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/terraform/etc/consul.d/consul.service]: I think it should be ok. Pushed that and will test.

File: e2e/terraform/README.md [hunk 3]
```diff
@@ -40,66 +66,25 @@ terraform apply
 > where the ssh service isn't yet ready. That's ok and expected; they'll get
 > retried. In particular, Windows instances can take a few minutes before ssh
 > is ready.
+>
+> Also note: When ACLs are being bootstrapped, you may see "No cluster
+> leader" in the output several times while the ACL bootstrap script
+> polls the cluster to start and and elect a leader.
 
 ## Nomad Version
```
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/terraform/README.md]: Fixed!

File: e2e/terraform/hcp_consul.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/terraform/hcp_consul.tf]: Seems to work

File: e2e/terraform/etc/consul.d/consul.service (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/terraform/etc/consul.d/consul.service]: Done
